DataSource/Cloudwatch: Add Auth Option and Cache Config Per DataSource

patstrom opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

This PR does two main things. 

## First thing: Authentication
It refactors the entire tsdb/cloudwatch authentication flow and adds a `AWS SDK Default` authentication option that does what one would expect. All other options are refactored to only perform *exactly* what is specified and thus any implicit behaviour is removed.

The details of this are isolated to the first commit of this PR (8db20e4f31) and it contains some more information.

## Second thing:  No more global caches
The current cloudwatch implementation makes frequent use of global caches, both for credentials (or sessions after 8db20e4f31) and for custom metrics and dimensions. This is obviously a problem when someone for example adds two cloudwatch datasources or when custom metrics are requested for two different regions at the same time. To be honest I haven't really bothered to find out all the ways this can cause a problem but caching everything on a per-datasource basis would solve them, as well as make more sense conceptually.

The changes are in commit c37277eac6 and you can find more information there

**Which issue(s) this PR fixes**:

This PR is the "proper" solution mentioned [here](https://github.com/grafana/grafana/issues/20473#issuecomment-559181587) for #20473.

While https://github.com/grafana/grafana/pull/21594 does solve this specifically for EKS IAM Roles for Service Accounts the new `AWS SDK Default` (which just uses the default chain as specified by the SDK) is more future-proof in the sense that any new feature can be taken advantage of by simply updating the dependency rather than replicating it.

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->


**Special notes for your reviewer**:

*I need help testing this*! 

This MR could have wide implications for people relying on these potential bugs and implicit behavior. I do doubt anyone is actually relying on the bugs but I think *a lot* of user will have to update their Cloudwatch authentication used. This can of course be frustrating but personally I find it even more frustrating that sometimes things are implicit and sometimes they are not. In order to set up Grafana in an EKS cluster myself the first time I didn't want any of the available plugins (I wanted SDK default) so I just chose one and provided bad options until it eventually used some default credential as fallback.


I've tested this locally with all the different options. Credentials file and static credentials will *only* choose their respective providers so no implicit behavior at all there, since it works locally it should work everywhere. I still need to make sure the AWS SDK Default and Assume Role (which uses the sdk default for AssumeRole credentials) works in the following cases :

- [x] EC2 Instance with assigned role
- [x] Docker inside EC2 Instance with assigned role
- [x] ECS Task using the `aws_container_credentials_relative_uri` environment variable
- [x] EKS Using Web Identity

I also need suggestions/ideas for allowing the interfaces returned by the `getXClient` to be mocked for testing purposes. Right now the `metric_find_query_test.go` doesn't work.

patstrom committed:
cloudwatch: Add AWS SDK Default Auth Provider

A recurring problem seems to be that the cloudwatch credentials provider
is manually specified and is therefore lagging behind the features AWS
releases. The `remoteCredsProvider` seems to be a workaround for this in
the case of ECS, where official support was added in 2016. With the new
web identity provider (which is the go-to, officially recommended way to
authenticate in an EKS cluster) this is a problem once again.

The solution I came up with is to add an Auth Provider that is the
default aws sdk credentials chain. This option should be the most
intuitive for people running Grafana in a managed service such as ECS or
EKS and will allow Grafana to better align with how many other projects
authenticate to AWS.

The aws sdk go caches credentials by default so I started with removing
the manual cache. I replaced this cache with a session cache instead.
This caches sessions per region and invalidates them if the
authentication type is updated. By caching a session instead of
credentials we can avoid loading the configuration options each time.
By invalidating them when the authentication type changes we make our
intent more explicit and eliminate any unexpected behaviour
that the old credentials cache could cause when updating datasources
with cached credentials.

The credentials cache did not do much explicit handling of the AuthType
provided. In fact, the credentials chain it created (and cached) ended
up being a mix of all the different types depending on what you had
specified. It also included some seemingly arbitrary chosen defaults.
The current solution is very explicit and does *only* what you tell it
to. Since some of this implicit behaviour might have been relied upon
this might be a breaking change.

The access patterns were also differing between the CloudWatchAPI and
the ec2/rgta APIs so I unified those. Previously the CloudWatchAPI
client was accessed via a `getClient` function whereas the other two
services were struct fields and required `ensureClient` to be called
before using them. Since the latter behaviour was not particularly
thread safe and would likely cause unexpected errors if someone called
it with something other than `"default"` as parameter I decided on
standardizing on the former method.

patstrom committed:
cloudwatch: Cache information per datasource

The cloudwatch package has greatly utilized global caches for both
credentials and custom metrics and custom dimensions. There are a number
of problems with this.

By caching credentials on a global basis, adding two cloudwatch
datasources can cause a lot of contention and they can end up using each
others credentials. Likewise when caching custom metrics and dimensions.
The caches themselves weren't necesarily invalidated when the datasource
was updated either, which can lead to confusion and unexpected
behaviour. All in all, the cloudwatch package likely suffered from many
concurrency bug due to using global configuration caches when they
should've been per data source.

Previously, a CloudWatchExecutor was created for each request rather
than for each datasource. While this is not a problem in itself it makes
moving the caches to a per-datasource basis more difficult. This patch
will keep a global cache of executors, where one executor is kept for
each data source id. This executor will then itself cache sessions and
custom metrics and dimension on at least a per-region basis.

Keeping a CloudWatchExecutor for each datasource also allows us to
actually cache the datasource information itself as well rather than
assigning it each time which was the previous behaviour.

An entry in the executor cache is considered invalid if it is missing
(obviously) or if the requested datasource has been updated since we
cached it. This check is made by using the Version field and relies on
it being monotonically increasing. If the cached executor has a version
that is greater than or equal to the version of the specified datasource
then we return the cached one, otherwise we create a new one
invalidating all the current caches.

patstrom commented:
I did the `make build-docker-full` and uploaded it as `patstrom/grafana:c37277eac68a314fd49f6382ca6e868e48771aa8` for testing purposes

patstrom commented:
I've tested both a "raw" Docker (running on amazon linux) container and in an EKS cluster and both works as expected in the preliminary testing I've done. Which is making sure it can authenticate and query for some data in multiple regions. Haven't tested custom metrics and dimensions.

patstrom commented:
I've tested all the different configuration now (Raw EC2, Container on raw EC2, ECS and EKS) and the `AWS SDK Default` option and the `ARN` option (using the `AWS SDK Default` option to assume the role) works as expected. The sessions are properly cached between queries and datasources as well. Doing a query in a new region creates a new session and caches it. Creating a new datasource creates a new executor with it's own set of sessions.

I don't really know how to test custom metrics since I don't really know how they work in detail. However, since it works for sessions I assume it kind of works for metrics as well. Especially since the actual implementation changes are very minimal (move the map to the executor and reference that instead of using the global one).

Something to consider is that right now (both before and after this patch) the error reporting is fairly lacking. It just says that it's not working and returns a `500` error and the user will have to check the logs. It might be worth it to at least create an issue to at least differentiate between unauthorized and something is actually wrong.

@ryantxu @aknuds1

patstrom committed:
Cloudwatch: Use tooltips to clarify authentication behavior

patstrom commented:
@ryantxu @aknuds1 could you at least confirm this might make it in 7.0? If it's not something you might want to merge anyway I won't bother fixing the unit tests.

aknuds1 commented:
Sorry, I haven't been able to check this one yet. I will try tomorrow.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
I merged with Grafana master in order to resolve conflicts. Trying to review it today.

aknuds1 committed:
Fix vendoring

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

aknuds1 commented:
Are you able to fix the CI errors?

aknuds1 commented on pkg/tsdb/cloudwatch/metric_find_query.go:
```diff
@@ -759,80 +710,80 @@ func getAllMetrics(cwData *DatasourceInfo) (cloudwatch.ListMetricsOutput, error)
-		customMetricsMetricsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace] = &CustomMetricsCache{}
-		customMetricsMetricsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace].Cache = make([]string, 0)
+	if _, ok := e.customMetricsMetricsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace]; !ok {
+		e.customMetricsMetricsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace] = &CustomMetricsCache{}
+		e.customMetricsMetricsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace].Cache = make([]string, 0)
```
```suggestion
		e.customMetricsMetricsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace] = &CustomMetricsCache{
		        Cache: make([]string, 0),
		}
```

aknuds1 commented on pkg/tsdb/cloudwatch/metric_find_query.go:
```diff
@@ -759,80 +710,80 @@ func getAllMetrics(cwData *DatasourceInfo) (cloudwatch.ListMetricsOutput, error)
-		customMetricsDimensionsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace] = &CustomMetricsCache{}
-		customMetricsDimensionsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace].Cache = make([]string, 0)
+	if _, ok := e.customMetricsDimensionsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace]; !ok {
+		e.customMetricsDimensionsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace] = &CustomMetricsCache{}
+		e.customMetricsDimensionsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace].Cache = make([]string, 0)
```
```suggestion
		e.customMetricsDimensionsMap[dsInfo.Profile][dsInfo.Region][dsInfo.Namespace] = &CustomMetricsCache{
		        Cache: make([]string, 0),
		}
```

aknuds1 commented on pkg/tsdb/cloudwatch/session.go:
```diff
@@ -0,0 +1,94 @@
+	case "arn":
+		stsSess, err := session.NewSession(regionConfiguration)
+		if err != nil {
+			return nil, fmt.Errorf("creating session for AssumeRoleProvider: %w", err)
```
```suggestion
			return nil, fmt.Errorf("creating session for AssumeRoleProvider failed: %w", err)
```

aknuds1 commented on pkg/tsdb/cloudwatch/session.go:
```diff
@@ -0,0 +1,94 @@
+
+	sess, err := s.newAwsSession(dsInfo)
+	if err != nil {
+		return nil, fmt.Errorf("creating new session for region %q: %w", region, err)
```
```suggestion
		return nil, fmt.Errorf("creating new session for region %q failed: %w", region, err)
```

aknuds1 commented on public/app/plugins/datasource/cloudwatch/components/ConfigEditor.tsx:
```diff
@@ -231,7 +237,10 @@ export class ConfigEditor extends PureComponent<Props, State> {
-                <FormLabel className="width-14" tooltip="ARN of Assume Role">
+                <FormLabel
+                  className="width-14"
+                  tooltip="ARN of Role to Assume. The AWS SDK Default credentials-chain will be used to assume it."
```
```suggestion
                  tooltip="ARN of role to assume. The AWS SDK Default credentials-chain will be used to assume it."
```

aknuds1 reviewed (changes requested):
Please see my suggestions :)

patstrom committed:
Implement PR suggestions

patstrom commented on pkg/tsdb/cloudwatch/cloudwatch.go:
```diff
@@ -2,22 +2,67 @@ package cloudwatch
+	sessions *sessionCache
+
+	// We cache custom metrics and dimensions on a per-datasource per-version basis
+	// These are of type (profile -> region -> namespace
```
No the missing closing parentheses was a typo! I usually write out the (intended) "type" of the maps in such a fashion if they use built-in types for keys.

patstrom committed:
cloudwatch: Create interface for AWS API Clients

Instead of keeping a sessionCache of AWS SDK Go sessions per-region in
the CloudWatchExecutor it will now instead require an interface that
implement a function for fetching each respective client. This also
means that the executor itself does not implemented any methods for
fetching the clients.

The existing sessionCache implements this interface and the end result
is that responsibility of generating the different API clients now lies
entirely with the sessionCache, rather than the executor using the
sessionCache for configuration and creating clients on its own.

Essentially, this creates an abstraction around the entire AWS API
(since the interface functions themselves returns interfaces) and
allows users to create a CloudWatchExecutor that returns mocked clients
(and thus mocked API responses) for testing purposes.

papagian commented:
@patstrom 
thank you for contribution! your work seems interesting!

Since, as you mentioned, these modifications can have implications for people using the plugin we would prefer to merge it after 7.0 in order to have the time to evaluate it better.

Also, since you basically introduce two different features, spitting it in two separate pull requests would help the review process and increase the chances from them to be accepted and merged.

aknuds1 commented:
If feasible, I would agree with @papagian it's preferable to split into multiple PRs (with the caveat that I haven't tried to understand the PR in depth, I lack experience with our CW integration).

patstrom committed:
cloudwatch: Update ConfigEditor snapshots

patstrom commented:
When you say after 7.0 does that mean this would be included in the 8.0 release?

Honestly dividing this up into two separate PRs isn't really feasible. 

The main point of this PR is to move up the abstraction ladder for the AWS SDK. The functionality before this is that a `credentials.Credentials` is cached globally for each `authType:AccessKey:Profile:AssumeRoleARN` combination, which is an unintuitive and incorrect way to handle this. The `cacheKey` is created with the assumption that if two datasources has the same authentication type, access key, profile and role to assume they should have the same credentials. Which I suppose is good enough in the majority of cases but it's not *correct*.

Another problem with caching the credentials is also that it requires the creation of a `session.Session` for every single request. Creating a session will load configuration files and whatnot and AWS [recommends caching sessions when possible](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/sessions.html). So the "problem" then is that the session includes all other necessary configuration (like which region it's supposed to connect to) and thus the cache must be per-region. So in order to move a session with a global cache the already flawed `cacheKey` must include the region. Dividing up these two features (one of which I consider more of a bug than a feature) is thus not really feasible.

The custom metrics cache is literally just moved from a global namespace to an instance of the `CloudWatchExecutor` struct. Implementation-wise I've done nothing with it and honestly after this PR it could probably be simplified a lot.

Since I've also updated the executor's own interface for fetching these clients it's also difficult to divide this up and still have tests passing and not too many eventual merge conflicts.

I have provided somewhat detailed commit messages and  there is quite some information in the linked issue and PR for the motivation behind this.

Now that I've gotten to know the Cloudwatch implementation a little better I could go through and add some more inline documentation about how the different components work and interact with each other.

aknuds1 commented:
Thanks for clarifying @patstrom - I will try to give it a more in-depth review next week.

aknuds1 commented:
Having a look :) I finally have time after getting v7 beta out the door.

aknuds1 commented:
@patstrom There are some pretty heavy merge conflicts due to recent Cloudwatch work, could you please reconcile them? I looked at the conflicts, and I don't know how to solve them.

patstrom committed:
Merge branch 'master' into f/add-aws-sdk-default

patstrom committed:
Remove vendor

patstrom committed:
Add CLOUDWATCH_DEFAULT_REGION constant

patstrom committed:
Don't pass around API clients

The CloudWatchExecutor already keeps tracks of aws sdk go service
clients so there is no need to pass those around as function parameters
on a method of CloudWatchExecutor.

Test can instead be implemented by providing the mockClients
implementation of clientsCache to the executor. A mockedLogs client
should also not have hard-coded responses as that would make writing
new tests more difficult.

patstrom commented:
I've now resolved all the conflicts with master!

While doing so I also refactored the CloudWatch logs implementation to follow the pattern for keeping track of aws-sdk-go clients established by this PR. I also refactored to tests so that the mock implementation of the CloudWatchLogsAPI doesn't have hard-coded responses (which makes writing new tests very cumbersome).

What is the idea here? To add it to the 7.0 beta? Or hold it off until 8.0 (since while the break is minor it is still technically a breaking change). Because I imagine 8.0 isn't due until at least 6 months and there will probably be a lot more conflicts happening up until then...

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

haghabozorgi commented:
just wanted to see if there has been any movement on this pr? we are currently using docker container built from patstrom's branch (thanks @patstrom for your efforts) for 'iam roles for service accounts'   support and would like to switch back to upstream image when feasible.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

aknuds1 commented:
@patstrom We are trying to review it as quickly as possible, but I've been held up with releasing Grafana 7 betas (3 so far) unfortunately. I'm diving into your code now :)

aknuds1 committed:
Fix stylistic issues

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
@patstrom As I review, I'm going to make some stylistic fixes (for example ones that will be caught with stricter linting, as we plan to enable). Hope you don't mind.

aknuds1 committed:
Pass CW logs API client down

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Move test type to test file

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
@patstrom Would you be able to give me some instructions on how to test this PR (manually)? I'm new to our CloudWatch functionality, and our main CW expert is off so I have to brush up :) I'm also open to discussing this on the Grafana Community slack if you want to ping me there.

patstrom commented:
The way I tested it manually was that I set up a "raw" ec2 instance, an ECS cluster and a EKS cluster and then for each of them I checked that the authentication providers did exactly what they are supposed to do. I.e. when providing access key id and secret it *only* uses those (and nothing else configured on the server), when using the credentials file option it *only* uses that option and that it can find everything appropriate (i.e. any attached IAM roles) when using the AWS SDK Default and ARN providers.

If I recall correctly I created two IAM roles, one that can access cloudwatch and one that can assume the one with access. Then I attached the role with cloudwatch access to a EC2 Instance, a ECS Cluster and a kubernetes service account in a EKS-managed cluster and made sure all of them could find the credentials (I also had some debug prints to make sure the cache worked correctly) and access cloudwatch metrics. Then I switched out the role to the one that can assume the cloudwatch role and made sure the Assume Role provider worked (by assuming the cloudwatch role). It's a little tricky to make sure that the trust policies for the IAM roles are correct in both cases (you have to switch them out when testing the other provider) but that's essentially all I did.

aknuds1 commented:
@patstrom Are you on the community Slack?

patstrom commented:
If it's the same as the Loki one I might be. Give me a minute

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

patstrom committed:
cloudwatch: Update authentication documentation

Write a little more about the different authentication providers and
clarify their behaviour.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

aknuds1 committed:
Merge branch 'f/add-aws-sdk-default' of github.com:patstrom/grafana into f/add-aws-sdk-default

oddlittlebird commented on docs/sources/features/datasources/cloudwatch.md:
```diff
@@ -30,9 +30,7 @@ Grafana ships with built in support for CloudWatch. You just have to add it as a
-| _Auth Provider_            | Specify the provider to get credentials.                                                                |
-| _Credentials_ profile name | Specify the name of the profile to use (if you use `~/.aws/credentials` file), leave blank for default. |
-| _Assume Role Arn_          | Specify the ARN of the role to assume                                                                   |
+| _Authentication Provider_  | Specify which method Grafana will use to find the AWS credentials.                                      |
```
```suggestion
| _Authentication Provider_  | Specify which method Grafana uses to find the AWS credentials.                                      |
```

oddlittlebird commented on docs/sources/features/datasources/cloudwatch.md:
```diff
@@ -88,14 +86,33 @@ Here is a minimal policy example:
-- IAM role for Amazon EC2.
-
-See the AWS documentation on [Configuring the AWS SDK for Go](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html)
+There are four different authentication providers available. The `AWS SDK
```
These line breaks are painful. Please fix them so this is easier to read.

oddlittlebird commented on pkg/tsdb/cloudwatch/cloudwatch.go:
```diff
@@ -9,25 +9,48 @@ import (
+	executorLock = sync.Mutex{}
+)
+
+// CloudWatchExecutor represents a struct holding enough information to execute
```
```suggestion
// CloudWatchExecutor represents a structure holding enough information to execute
```

oddlittlebird commented on pkg/tsdb/cloudwatch/cloudwatch.go:
```diff
@@ -9,25 +9,48 @@ import (
+
+// CloudWatchExecutor represents a struct holding enough information to execute
+// CloudWatch queries for a specific datasource. It caches AWS SDK Sessions on
+// a regional basis for each datasource version in order to load configuration as
```
```suggestion
// a regional basis for each data source version in order to load configuration as
```

oddlittlebird replied:
"data source" is always two words.

oddlittlebird commented on docs/sources/features/datasources/cloudwatch.md:
```diff
@@ -88,14 +86,33 @@ Here is a minimal policy example:
+The `Access & secret key` corresponds to the
+[StaticProvider](https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/#StaticProvider)
+and will use the given access key id and the secret key to authenticate. If
+this provider fails for whatever reason the authentication will fail. No
```
```suggestion
this provider fails for whatever reason, then the authentication will fail. No
```

oddlittlebird commented on docs/sources/features/datasources/cloudwatch.md:
```diff
@@ -88,14 +86,33 @@ Here is a minimal policy example:
+credentials file (on account of the SharedCredentialsProvider being part of the
+default chain) this option allows you to specify which profile to use without
+messing with environment variables. This option will not do any implicit
+fallbacks to other providers. If the SharedCredentialsProvider fails
```
```suggestion
fallbacks to other providers. If the SharedCredentialsProvider fails, then
```

patstrom committed:
Change "will use" to "uses"

patstrom committed:
Adjust line breaks to have longer lines

patstrom committed:
struct -> structure

patstrom committed:
Data source is two words

patstrom committed:
Reword sentence about authentication failure

patstrom commented:
Thank you for the review, @oddlittlebird! I've implemented your suggested changes.

I just now realize I could've committed them directly here on GitHub which (I assume?) would've credited you, sorry about that.

I don't know if it's up to me or you to mark them as "resolved".

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

oddlittlebird commented:
In my PRs, I usually mark them as Resolved when I resolve them, but I don't think it makes a big deal.

Your next step is to re-request my review. See the dropdown arrow next to "1 change requested"? Click that and you will see the option.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

oddlittlebird reviewed (approved):
LGTM

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Fix previous merge

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into f/add-aws-sdk-default

aknuds1 committed:
CloudWatch: Simplify

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Fix issues from merge

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

hugohaggmark commented on public/app/plugins/datasource/cloudwatch/components/ConfigEditor.tsx:
```diff
@@ -227,7 +233,10 @@ export class ConfigEditor extends PureComponent<Props, State> {
+                  className="width-14"
+                  tooltip="ARN of role to assume. The AWS SDK Default credentials-chain will be used to assume it."
+                >
                   Assume Role ARN
```
I'm not super familiar with this code so this might be way off, but wouldn't it make sense to use the `label` prop here (if we make the texts the same)?
```suggestion
                  {authProviderOptions[1].label}
```

patstrom replied:
It probably does :sweat_smile: 

I guess it should use the `label` for all options then?

hugohaggmark replied:
I think so as long as the texts align.

patstrom replied:
Oh wait I think I've gotten confused.

This label is for the textbox where you enter the actual ARN of the role to assume. The label of the authentication provider is something different.

You first select that you want Grafana to assume a specific role (`{authProviderOptions[1].label}`) and then you get a new field where you enter which role that is (`Assume Role ARN`).

Maybe it should be changed to something like `ARN of role to assume` but it probably shouldn't match the label of the Authentication Provider.

patstrom replied:
I think this is no longer relevant with the latest commit (dd79792). But there has been new frontend changes that I would be happy if you could review :)

hugohaggmark replied:
@patstrom sure I'll re-review it again on Monday, have a pleasant weekend!

hugohaggmark reviewed (commented):
The front-end part looks good, I found just a small nitpick.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

dgholz commented on public/app/plugins/datasource/cloudwatch/components/ConfigEditor.tsx:
```diff
@@ -227,7 +233,10 @@ export class ConfigEditor extends PureComponent<Props, State> {
-                <InlineFormLabel className="width-14" tooltip="ARN of Assume Role">
+                <InlineFormLabel
+                  className="width-14"
+                  tooltip="ARN of role to assume. The AWS SDK Default credentials-chain will be used to assume it."
```
Thanks for working on this! I'm a Grafana Cloud user and I'd like to be able to specify the credentials used when assuming the role, since the Grafana Cloud instances don't have IAM instance profiles (runs on GCP) and won't have default credentials configured (I don't have access to local disk/k8 to create `~/.aws/config` or to set env vars for AWS SDK).

Will it be possible to configure both a Role to assume & non-default creds to use when assuming it?

rgarrigue replied:
Same question, as a user I expect to be able to specify aws_access_key_id, aws_secret_access_key and assumed role ARN through UI or API.

patstrom replied:
This is something I did consider but in order to make it intuitive more frontend development was required than I am comfortable doing :sweat_smile:

Or at least in the way I originally thought about it. I just though of another solution that might work. I'll give it a go this in the coming days and see how it turns out.

patstrom replied:
I just committed a change that moves `Assume Role ARN` *from* the `Authentication Provider` dropdown into its own textbox. 

This would then mean that the given `Authentication Provider` is used to assume the specified role (if a role is specified in the first place).

Does this solve your concerns or have I missed something? :)

rgarrigue replied:
I think so ! Thanks for taking care of this :-)

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

patstrom committed:
Move Assume Role from a provider to a global option

The Assume Role authentication provider is not "really" a provider,
rather is just an option as to what to do with the specified provider.

This commit allows you to assume a role using any of the available
authentication providers by moving the `Assume Role ARN` option from the
Authentication Provider dropdown to a "global" option similar to the
default region.

This commit refactors the session.Session generation to first
instantiate an aws session in a manner consistent with the selected
authentication provider and then checking whether an `Assume Role ARN`
was given, in which case it will wrap the session in an
`AssumeRoleProvider`. This is in contrast to treating the `Assume Role
ARN` configuration as an authentication provider in itself.

This change will allow users to use e.g access key id and secret access
key specified in the UI/API to still assume a separate IAM role.

patstrom committed:
Remove arn as authentication type in provisioning documentation

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into patstrom-add-aws-sdk-default

patstrom committed:
Document pod security context for EKS

EKS IAM Roles for Service Accounts requires a security context in order
for Grafana to find the credentials. This commit adds a section that
describes how (and why) to configure your pod appropriately.

Credits to @FaHeymann who provided the correct answer when helping out
@ingshtrom in https://github.com/grafana/grafana/issues/20473#issuecomment-638268104

patstrom commented:
I've divided up this PR into 4 distinct ones (where each one is based on the last)

https://github.com/grafana/grafana/pull/25555
https://github.com/grafana/grafana/pull/25556
https://github.com/grafana/grafana/pull/25557
https://github.com/grafana/grafana/pull/25558

aknuds1 commented:
@patstrom Can we close this one since you divided it up into smaller ones?

aknuds1 commented:
Closing this since it's divided into smaller PRs, to avoid confusion.
